### PR TITLE
Switch to internalclientset

### PR DESCRIPTION
**Pull request by soltysh**
This is switching us to use internalclientset instead of plain client.

@ncdc we need this land before the rebase
@bparees for the build parts
@kargakis for the deployment parts

### CODE REVIEW

**Comment by soltysh** [comment]
[test]

File: pkg/deploy/api/test/ok.go [hunk 1]
```diff
@@ -22,7 +22,9 @@ const (
 func OkDeploymentConfig(version int64) *deployapi.DeploymentConfig {
 	return &deployapi.DeploymentConfig{
 		ObjectMeta: kapi.ObjectMeta{
-			Name: "config",
+			Name:      "config",
+			Namespace: kapi.NamespaceDefault,
```
Comment (kargakis) [inline pkg/deploy/api/test/ok.go]: Why do you need this change? It will probably break tests.

File: pkg/deploy/registry/deploylog/rest.go [hunk 2]
```diff
@@ -221,7 +225,7 @@ func (r *REST) returnApplicationPodName(target *kapi.ReplicationController) (str
 	selector := labels.Set(target.Spec.Selector).AsSelector()
 	sortBy := func(pods []*kapi.Pod) sort.Interface { return controller.ByLogging(pods) }
 
-	pod, _, err := kcmdutil.GetFirstPod(r.pn, target.Namespace, selector, r.timeout, sortBy)
+	pod, _, err := kcmdutil.GetFirstPod(r.oldPn, target.Namespace, selector, r.timeout, sortBy)
```
Comment (kargakis) [inline pkg/deploy/registry/deploylog/rest.go]: Can you open a follow-up upstream to switch this to use a clientset?

Comment (kargakis) [inline pkg/deploy/registry/deploylog/rest.go]: Or issue is just fine

Comment (kargakis) [inline pkg/deploy/registry/deploylog/rest.go]: cool, so we just need the rebase

File: pkg/deploy/api/test/ok.go (hunk 1 above)
Comment (soltysh) [inline pkg/deploy/api/test/ok.go]: `fake.NewSimpleClientset` requires to have namespace set, it panicked without it. The unit tests are green already ;)

File: pkg/deploy/registry/deploylog/rest.go (hunk 2 above)
Comment (soltysh) [inline pkg/deploy/registry/deploylog/rest.go]: That's why this PR is a prereq :stuck_out_tongue_winking_eye:

**Comment by ncdc** [comment]
cc @deads2k @liggitt @sttts

File: pkg/build/controller/image_change_controller_test.go [hunk 3]
```diff
@@ -305,7 +305,8 @@ func mockBuildConfig(baseImage, triggerImage, repoName, repoTag string) *buildap
 	dockerfile := "FROM foo"
 	return &buildapi.BuildConfig{
 		ObjectMeta: kapi.ObjectMeta{
-			Name: "testBuildCfg",
+			Name:      "testBuildCfg",
+			Namespace: kapi.NamespaceDefault,
```
Comment (bparees) [inline pkg/build/controller/image_change_controller_test.go]: is something requiring we set this namespace now?

**Comment by bparees** [comment]
I have concerns/questions about all the default namespacing you're introducing in the tests, otherwise lgtm.

File: pkg/build/generator/generator_test.go [hunk 4]
```diff
@@ -741,7 +741,7 @@ func TestGenerateBuildFromConfig(t *testing.T) {
 	bc := &buildapi.BuildConfig{
 		ObjectMeta: kapi.ObjectMeta{
 			Name:      "test-build-config",
-			Namespace: "test-namespace",
+			Namespace: kapi.NamespaceDefault,
```
Comment (deads2k) [inline pkg/build/generator/generator_test.go]: > why?

The new testing mock used by generated fake clients requires that namespaced resources come in with namespaces.  It doesn't understand defaulting.

Comment (bparees) [inline pkg/build/generator/generator_test.go]: and we don't consider that an upstream bug? ok.....

Comment (bparees) [inline pkg/build/generator/generator_test.go]: also that still doesn't explain why this object which had a valid namespace, is being changed to use a different namespace?

Comment (bparees) [inline pkg/build/generator/generator_test.go]: (I don't *think* this particular test actually cares what the namespace value is, but it seems odd to change it)

File: pkg/deploy/registry/deploylog/rest_test.go [hunk 5]
```diff
@@ -144,15 +146,49 @@ func mockREST(version, desired int64, status api.DeploymentStatus) *REST {
 				Phase: kapi.PodRunning,
 			},
 		}
-		fakePn.PrependReactor("get", "pods", func(action ktestclient.Action) (handled bool, ret runtime.Object, err error) {
+		oldPn.PrependReactor("get", "pods", func(action ktestclient.Action) (handled bool, ret runtime.Object, err error) {
```
Comment (sttts) [inline pkg/deploy/registry/deploylog/rest_test.go]: What's the reason for this pretty big change here?

File: pkg/build/controller/image_change_controller_test.go (hunk 3 above)
Comment (soltysh) [inline pkg/build/controller/image_change_controller_test.go]: Yes, the fake clientset is now namespace aware, without it you'll get nasty panic complaining that you're trying to use namespaced resource without namespace.

File: pkg/deploy/registry/deploylog/rest_test.go (hunk 5 above)
Comment (soltysh) [inline pkg/deploy/registry/deploylog/rest_test.go]: rolling and recreate invoke commands from `k8s.io/kubernetes/pkg/kubectl` package, which is using clientset only in 1.5, which we'll get after the rebase. Example line:
`kubectl.ScalerFor(kapi.Kind("ReplicationController"), oldClient)` for that reason we need old k8s client in only certain places, which will be removed after the rebase. See my discussion with Michalis in this PR.

**Comment by soltysh** [comment]
@ncdc I think this is ready, eventual leftovers we'll handle afterwards/during the rebase itself.

I need lots of eyes on your area: 
@knobunc for networking stuff
@bparees for build stuff
@deads2k @liggitt for auth stuff
@smarterclayton overall

**Comment by smarterclayton** [comment]
Are there any changes here that aren't just simple refactoring?

**Comment by ncdc** [comment]
@soltysh needs rebase. Last failure flaked on #11775 and this:

```
--- FAIL: TestPullThroughInsecure (4.14s)
	dockerregistry_pullthrough_test.go:175: External registry got GET /v2/
	dockerregistry_pullthrough_test.go:175: External registry got GET /v2/test/testrepo/manifests/testtag
	dockerregistry_pullthrough_test.go:175: External registry got HEAD /v2/test/testrepo/manifests/testtag
	dockerregistry_pullthrough_test.go:175: External registry got GET /v2/test/testrepo/manifests/sha256:958608f8ecc1dc62c93b6c610f3a834dae4220c9642e6e8b4e0f2b3ad7cbd238
	dockerregistry_pullthrough_test.go:246: unexpected status 0: unversioned.Status{TypeMeta:unversioned.TypeMeta{Kind:"", APIVersion:""}, ListMeta:unversioned.ListMeta{SelfLink:"", ResourceVersion:""}, Status:"Failure", Message:"Internal error occurred: invalid character '\\n' in string literal", Reason:"InternalError", Details:(*unversioned.StatusDetails)(0xc825c3b360), Code:500}
```

**Comment by soltysh** [comment]
> Are there any changes here that aren't just simple refactoring?

Most is just simple refactoring, but there are a few places where bigger changes were required.

**Comment by soltysh** [comment]
Rebased and fixed the test Andy pointed out.

File: pkg/cmd/cli/describe/projectstatus_test.go [hunk 6]
```diff
@@ -214,8 +213,8 @@ func TestProjectStatus(t *testing.T) {
 			Contains: []string{
 				// this makes sure that status knows this can push.  If it fails, there's a "(can't push image)" next to like #8
 				" hours\n  build #7",
-				"on fedora:23",
-				"-> repo-base:latest",
+				"on istag/fedora:23",
+				"-> istag/repo-base:latest",
```
Comment (liggitt) [inline pkg/cmd/cli/describe/projectstatus_test.go]: why did this output change?

File: pkg/cmd/cli/describe/projectstatus_test.go [hunk 7]
```diff
@@ -313,9 +312,9 @@ func TestProjectStatus(t *testing.T) {
 			ErrFn: func(err error) bool { return err == nil },
 			Contains: []string{
 				"In project example on server https://example.com:8443\n",
-				"svc/galera[default] (headless):3306",
+				"svc/galera (headless):3306",
```
Comment (liggitt) [inline pkg/cmd/cli/describe/projectstatus_test.go]: same question

File: pkg/cmd/cli/describe/projectstatus_test.go (hunk 6 above)
Comment (soltysh) [inline pkg/cmd/cli/describe/projectstatus_test.go]: I was actually wondering why it wasn't like that before, looking at other test cases.

File: pkg/cmd/cli/describe/projectstatus_test.go (hunk 7 above)
Comment (soltysh) [inline pkg/cmd/cli/describe/projectstatus_test.go]: Here, situation is a bit more complicated. Fake clientset forces to always use namesapces, previously it was not required. I've added them and previously we've had parts of the petset created in default namespace and now it's all in one. Thus the change.

**Comment by smarterclayton** [comment]
I asked because the mechanical transform need to be separate from the
important changes somehow so we can focus on what needs reviewing.

On Nov 18, 2016, at 5:31 PM, Maciej Szulik <notifications@github.com> wrote:

*@soltysh* commented on this pull request.
------------------------------

In pkg/cmd/cli/describe/projectstatus_test.go
<https://github.com/openshift/origin/pull/11916>:

> @@ -313,9 +312,9 @@ func TestProjectStatus(t *testing.T) {
 			ErrFn: func(err error) bool { return err == nil },
 			Contains: []string{
 				"In project example on server https://example.com:8443\n",
-				"svc/galera[default] (headless):3306",
+				"svc/galera (headless):3306",

Here, situation is a bit more complicated. Fake clientset forces to always
use namesapces, previously it was not required. I've added them and
previously we've had parts of the petset created in default namespace and
now it's all in one. Thus the change.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/openshift/origin/pull/11916>, or mute the thread

**Comment by soltysh** [comment]
> I asked because the mechanical transform need to be separate from the important changes somehow so we can focus on what needs reviewing.

OK, I'll update that PR into what needs review and the rest.

**Comment by soltysh** [comment]
Looks like one another error:
```
--- FAIL: TestDeployScale (4.23s)
	deploy_scale_test.go:48: Couldn't create DeploymentConfig: the namespace of the provided object does not match the namespace sent on the request &api.DeploymentConfig{TypeMeta:unversioned.TypeMeta{Kind:"", APIVersion:""}, ObjectMeta:api.ObjectMeta{Name:"config", GenerateName:"", Namespace:"default", SelfLink:"/oapi/v1/namespaces/default/deploymentconfig/config", UID:"", ResourceVersion:"", Generation:0, CreationTimestamp:unversioned.Time{Time:time.Time{sec:0, nsec:0, loc:(*time.Location)(nil)}}, DeletionTimestamp:(*unversioned.Time)(nil), DeletionGracePeriodSeconds:(*int64)(nil), Labels:map[string]string(nil), Annotations:map[string]string(nil), OwnerReferences:[]api.OwnerReference(nil), Finalizers:[]string(nil), ClusterName:""}, Spec:api.DeploymentConfigSpec{Strategy:api.DeploymentStrategy{Type:"Recreate", CustomParams:(*api.CustomDeploymentStrategyParams)(nil), RecreateParams:(*api.RecreateDeploymentStrategyParams)(0xc82ffda3c0), RollingParams:(*api.RollingDeploymentStrategyParams)(nil), Resources:api.ResourceRequirements{Limits:api.ResourceList{"cpu":resource.Quantity{i:resource.int64Amount{value:10, scale:0}, d:resource.infDecAmount{Dec:(*inf.Dec)(nil)}, s:"10", Format:"DecimalSI"}, "memory":resource.Quantity{i:resource.int64Amount{value:10, scale:9}, d:resource.infDecAmount{Dec:(*inf.Dec)(nil)}, s:"10G", Format:"DecimalSI"}}, Requests:api.ResourceList(nil)}, Labels:map[string]string(nil), Annotations:map[string]string(nil)}, MinReadySeconds:0, Triggers:[]api.DeploymentTriggerPolicy{}, Replicas:1, RevisionHistoryLimit:(*int32)(nil), Test:false, Paused:false, Selector:map[string]string{"a":"b"}, Template:(*api.PodTemplateSpec)(0xc82d4cfa40)}, Status:api.DeploymentConfigStatus{LatestVersion:0, ObservedGeneration:0, Replicas:0, UpdatedReplicas:0, AvailableReplicas:0, UnavailableReplicas:0, Details:(*api.DeploymentDetails)(nil), Conditions:[]api.DeploymentCondition(nil)}}
	etcd.go:99: dumping etcd to "/tmp/openshift/test-integration//etcd-dump-runtime.call32.json"
FAIL
```
but I think I'll leave that for tomorrow to handle.

**Comment by soltysh** [comment]
Fixed the test and re-organized into 2 commits boring and interesting, per Clayton request. 

@smarterclayton, @liggitt @ncdc ptal

**Comment by soltysh** [comment]
Fixed all integration tests.

**Comment by soltysh** [comment]
Flake https://github.com/openshift/origin/issues/11016

[test]

**Comment by soltysh** [comment]
Flake https://github.com/openshift/origin/issues/10988

[test]

**Comment by soltysh** [comment]
Flake #11747

[test]

**Comment by soltysh** [comment]
Flake #10988

[test]

**Comment by soltysh** [comment]
Flake #11016

[test]

**Comment by soltysh** [comment]
Flake #11016

[test]

**Comment by soltysh** [comment]
Flake https://github.com/openshift/origin/issues/11775

[test]

**Comment by soltysh** [comment]
@smarterclayton finally green, yay! Are we good to merge this?

File: pkg/client/testclient/testclient.go [hunk 8]
```diff
@@ -1,20 +1,75 @@
...
 
 // NewFixtureClients returns mocks of the OpenShift and Kubernetes clients
-func NewFixtureClients(o testclient.ObjectRetriever) (osclient.Interface, kclient.Interface) {
-	oc := &Fake{}
-	oc.AddReactor("*", "*", testclient.ObjectReaction(o, registered.RESTMapper()))
-
-	kc := &testclient.Fake{}
-	kc.AddReactor("*", "*", testclient.ObjectReaction(o, registered.RESTMapper()))
+// with data populated from provided path.
+func NewFixtureClients(decoder runtime.Decoder, namespace, path string, objects ...runtime.Object) (client.Interface, kclientset.Interface, kclient.Interface, error) {
```
Comment (liggitt) [inline pkg/client/testclient/testclient.go]: this is a weird function now... it reads files from disk, it defaults namespaces... seems like three functions mixed up in one

Comment (soltysh) [inline pkg/client/testclient/testclient.go]: This was just keeping old functionality, what are you proposing?

**Comment by soltysh** [comment]
Rebased and fixed the testclient.go manual exception. The last thing brought up by @liggitt was wrt `NewFixtureClients` in `pkg/client/testclient/testclient.go` which does "reads files from disk, it defaults namespaces... seems like three functions mixed up in one" but all I did was maintain previous functionality not available in the clientset. @smarterclayton @liggitt are we good with this or should that be fixed/split into separate methods, but that will require rewriting those tests, as well.

**Comment by soltysh** [comment]
Rebased and fixed the `testclient.go` @liggitt ptal

**Comment by liggitt** [comment]
@soltysh ReadObjectsFromPath change LGTM, thanks

**Comment by soltysh** [comment]
Thx, @smarterclayton anything more or we're good to merge this as is?

**Comment by soltysh** [comment]
Rebased.

**Comment by soltysh** [comment]
Flake #10988

[test]

**Comment by soltysh** [comment]
Flake https://github.com/openshift/origin/issues/10951

[test]

**Comment by soltysh** [comment]
Flake #10988

[test]

**Comment by soltysh** [comment]
Flake https://github.com/openshift/origin/issues/10663

[test]

**Comment by soltysh** [comment]
Rebased.

**Comment by soltysh** [comment]
Flake #11016

[test]

**Comment by soltysh** [comment]
@smarterclayton anything more or we're good to go as is?

**Comment by smarterclayton** [comment]
Lgtm [merge]

**Comment by soltysh** [comment]
Flake #8502.

[merge]

**Comment by soltysh** [comment]
Flake #10951

[merge]

**Pull request merged**